src/stirling/source_connectors/socket_tracer/protocols/dns/stitcher.cc

```diff
@@ -240,6 +243,96 @@ RecordsWithErrorCount<Record> StitchFrames(std::deque<Frame>* req_frames,
+
+    // Search for matching req frame
+    for (auto& req_frame : *req_frames) {
+      // If the request timestamp is after the response, then it can't be the match.
+      // Nor can any subsequent requests either, so stop searching.
+
+      if (req_frame.timestamp_ns > resp_frame.timestamp_ns) {
+        break;
+      }
+
+      if (resp_frame.header.txid == req_frame.header.txid) {
+        StatusOr<Record> record_status = ProcessReqRespPair(req_frame, resp_frame);
+        if (record_status.ok()) {
+          entries.push_back(record_status.ConsumeValueOrDie());
+        } else {
+          VLOG(1) << record_status.ToString();
+          ++error_count;
+        }
+
+        // Found a match, so remove both request and response.
+        // We don't remove request frames on the fly, however,
+        // because it could otherwise cause unnecessary churn/copying in the deque.
+        // This is due to the fact that responses can come out-of-order.
+        // Just mark the request as consumed, and clean-up when they reach the head of the queue.
+        // Note that responses are always head-processed, so they don't require this optimization.
+        found_match = true;
+        req_frame.consumed = true;
+        break;
+      }
+    }
+
+    if (!found_match) {
+      VLOG(1) << absl::Substitute("Did not find a request matching the response. TXID = $0",
+                                  resp_frame.header.txid);
+      ++error_count;
+    }
+
+    // Clean-up consumed frames at the head.
+    // Do this inside the resp loop to aggressively clean-out req_frames whenever a frame consumed.
+    // Should speed up the req_frames search for the next iteration.
+    auto it = req_frames->begin();
+    while (it != req_frames->end()) {
+      if (!(*it).consumed) {
+        break;
+      }
+      it++;
+    }
+    req_frames->erase(req_frames->begin(), it);
+
+    // TODO(oazizi): Consider removing requests that are too old, otherwise a lost response can mean
+    // the are never processed. This would result in a memory leak until the more drastic connection
+    // tracker clean-up mechanisms kick in.
+  }
+
+  resp_frames->clear();
+
+  // After the external loop's lifecycle comes to an end we end up with the request deque
+  // having only those request frames which have not been consumed yet i.e. consumed = false
+  // so essentially these are the requests which could not be matched with any response frame.
+  // Hence we iterate over this request deque, add a default response to it, make a record and
+  // append those records at the end of the entries vector.
+  auto it = req_frames->begin();
+  while (it != req_frames->end()) {
+    if (!(*it).consumed) {
+      Frame default_resp_frame;
+      default_resp_frame.timestamp_ns = 99;
+      StatusOr<Record> record_status = ProcessReqRespPair(*it, default_resp_frame);
+      entries.push_back(record_status.ConsumeValueOrDie());
+    }
+    it++;
+  }
+  return {entries, error_count};
```

Comment:
This is the new part. A couple cases that we'll probably have to handle:
1) What if a request can't be matched with a response because the response is still in flight? The parser could run at any time, so we we have to be ready to handle that case. We might need to build in some assumptions about how delayed a response can be.

2) What's the rationale on `default_resp_frame.timestamp_ns = 99;`? We'll have to be careful with that because it could affect latency calculations.

Reply:
1. What should we consider as a timeout for the dns request. I think we can handle that using timestamp_ns.

2. In the case that we don't find a response for a request we're stitching a default ( empty ) response to pass req-resp pair to PrcoeessRequestResponsePair(). But we have a check on timestamp_ns so this timestamp needs to be some arbitrarily large number just to clear the stitcher tests.

Reply:
For (1), there's no set number, but nslookup uses a default of 2 seconds if I'm not mistaken. So that could be a starting point. Whatever we choose, it should definitely be a FLAG so we can change it later if required.

Reply:
timeout added in a3e9243ac0b0e094651eb59e2ed6252a01080658